File: clients/intellij-community/src/main/java/dev/cleanroommc/workbench/intellij/community/LocalReviewClient.java
package dev.cleanroommc.workbench.intellij.community;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import java.net.URI;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.function.BooleanSupplier;

/** Saved-source transport; no domain interpretation, mutation, or construction. */
final class LocalReviewClient {
    static List<String> arguments(String session, String baseline, String expected) {
        if (!session.matches("work-session-v2-[0-9a-f]{32}")) throw new IllegalArgumentException("Select one exact Work Session ID.");
        if (baseline.isEmpty() || baseline.length() > 1024 || baseline.indexOf('\0') >= 0
                || baseline.contains("\n") || baseline.contains("\r")) throw new IllegalArgumentException("Select an explicit local Git baseline.");
        var result = new ArrayList<>(List.of("context", "run", session, "--", "review", "local", "--baseline-ref=" + baseline));
        if (expected != null) {
            if (!expected.matches("source-review:sha256:[0-9a-f]{64}")) throw new IllegalArgumentException("Invalid review identity.");
            result.addAll(List.of("--expect-review", expected));
        }
        return List.copyOf(result);
    }

    static JsonObject validate(JsonObject envelope, Path workspace) throws Exception {
        JsonObject result = envelope.getAsJsonObject("result");
        Path root = workspace.toRealPath();
        if (!envelope.get("format").getAsString().equals("workbench-developer-action-v1")
                || envelope.get("exit_code").getAsInt() != 0
                || !Path.of(URI.create(envelope.getAsJsonObject("context").getAsJsonObject("selection").get("pack_uri").getAsString())).equals(root)
                || !result.get("format").getAsString().equals("workbench-local-source-review-v1")
                || !result.get("review_id").getAsString().matches("source-review:sha256:[0-9a-f]{64}")
                || !result.getAsJsonObject("baseline").get("revision").getAsString().matches("[0-9a-f]{40}([0-9a-f]{24})?")
                || !Path.of(URI.create(result.getAsJsonObject("baseline").get("root_uri").getAsString())).equals(root)
                || !Path.of(URI.create(result.getAsJsonObject("candidate").get("root_uri").getAsString())).equals(root)
                || !result.getAsJsonObject("authority").get("runtime_authority").getAsString().equals("none")
                || result.getAsJsonObject("authority").get("source_mutated").getAsBoolean()
                || result.getAsJsonObject("authority").get("construction_authorized").getAsBoolean()) {
            throw new IllegalArgumentException("Local review changed the selected workspace or source-only contract.");
        }
        for (String field : List.of("files", "changes", "findings", "checks", "relationships")) {
            if (!result.get(field).isJsonArray()) throw new IllegalArgumentException("Invalid local review collection.");
        }
        for (var value : result.getAsJsonArray("files")) {
            String path = value.getAsJsonObject().get("path").getAsString();
            if (path.isEmpty() || path.contains("\\") || path.contains(":") || path.indexOf('\0') >= 0 || path.contains("\n") || path.contains("\r")) {
                throw new IllegalArgumentException("Unsafe reviewed path.");
            }
            for (String part : path.split("/", -1)) if (part.isEmpty() || part.equals(".") || part.equals("..")) throw new IllegalArgumentException("Unsafe reviewed path.");
        }
        return result;
    }

    static JsonObject invoke(CoreLaunch launch, Path workspace, String session, String baseline, String expected, BooleanSupplier cancelled) throws Exception {
        if (!launch.host().equals("native")) throw new IllegalArgumentException("Local review currently requires a native Linux Workbench host.");
        String raw = CommandProcess.capture(launch, arguments(session, baseline, expected), 16 * 1024 * 1024, 120, workspace.toString(), cancelled);
        return validate(JsonParser.parseString(raw).getAsJsonObject(), workspace);
    }
}

File: clients/vscode/localReviewClient.js
"use strict";

const fs = require("node:fs");
const { pathToFileURL } = require("node:url");
const { invokeCoreJson } = require("./coreCommandClient");
const { resolveCoreLaunch } = require("./coreLaunch");

function reviewArguments(session, baseline, expected) {
  if (typeof session !== "string" || !/^work-session-v2-[0-9a-f]{32}$/.test(session)) {
    throw new Error("Select one exact Work Session ID.");
  }
  if (typeof baseline !== "string" || !baseline || baseline.length > 1024 || /[\0\r\n]/.test(baseline)) {
    throw new Error("Select an explicit local Git baseline.");
  }
  const args = ["context", "run", session, "--", "review", "local", `--baseline-ref=${baseline}`];
  if (expected !== undefined) {
    if (!/^source-review:sha256:[0-9a-f]{64}$/.test(expected)) throw new Error("Invalid review identity.");
    args.push("--expect-review", expected);
  }
  return args;
}

function validateReview(value, root) {
  const result = value?.result;
  if (value?.format !== "workbench-developer-action-v1" || value.exit_code !== 0
      || value.context?.selection?.pack_uri !== root || result?.candidate?.root_uri !== root
      || result?.baseline?.root_uri !== root || result?.format !== "workbench-local-source-review-v1"
      || !/^source-review:sha256:[0-9a-f]{64}$/.test(result.review_id)
      || !/^[0-9a-f]{40}([0-9a-f]{24})?$/.test(result.baseline.revision)
      || result.authority?.runtime_authority !== "none" || result.authority?.source_mutated !== false
      || result.authority?.construction_authorized !== false || !Array.isArray(result.files)
      || !Array.isArray(result.changes) || !Array.isArray(result.findings) || !Array.isArray(result.checks)) {
    throw new Error("Local review changed the selected workspace or source-only contract.");
  }
  for (const row of result.files) {
    if (typeof row.path !== "string" || /[\\:\0\r\n]/.test(row.path)
        || row.path.split("/").some((part) => !part || part === "." || part === "..")) throw new Error("Unsafe reviewed path.");
  }
  return result;
}

async function invokeLocalReview(executable, session, baseline, options = {}) {
  const launch = options.launch || resolveCoreLaunch(executable);
  if (launch.host !== "native") throw new Error("Local review currently requires a native Linux Workbench host.");
  const value = await invokeCoreJson(executable, reviewArguments(session, baseline, options.expected), {
    ...options, launch, maximumOutput: 16 * 1024 * 1024, timeoutMs: 120000, label: "Local source review",
  });
  return validateReview(value, pathToFileURL(fs.realpathSync(options.cwd)).href);
}

class ReviewEpoch {
  constructor() { this.sequence = 0; this.controller = undefined; }
  invalidate() { this.sequence += 1; this.controller?.abort(); this.controller = undefined; }
  begin() { this.invalidate(); this.controller = new AbortController(); return { sequence: this.sequence, signal: this.controller.signal }; }
  current(ticket) { return ticket.sequence === this.sequence && !ticket.signal.aborted; }
}

module.exports = { reviewArguments, validateReview, invokeLocalReview, ReviewEpoch };
